### DISCUSSION TITLE
Remove Tag From XML

### DISCUSSION THREAD

**Discussion by JJWest12** [nextgenhealthcare/connect#5076] [Q&A] [answered]
I am trying to build a channel take will take in some XML, and I want to remove a set of tags and everything within them. It seems really basic and I have been trying to mess around with getElementsByTagName and documentElement.removeChild based on some stuff I have found online but I cannot seem to get it working properly. Can anyone help out with this?

input: 

    <xml>
        <a>hi</a>
        <b>bye</b>
    </xml>

Desired output:

    <xml>
        <b>bye</b>
    </xml>

**Comment by pacmano1** [comment, marked as answer]
You need to provide more detail or is your XML really that simple?

Nodes can repeat  or be further embedded down in XML.

**Reply by JJWest12** [reply]
Going very basic to start so that is literally my example lol. I just need to figure out how to completely remove everything from a to /a in the message

**Reply by pacmano1** [reply]
`delete msg['a']` 

You said very basic.

**Reply by JJWest12** [reply]
HA! I had no idea you could just use the word "delete"! Thanks!

**Reply by pacmano1** [reply]
Remember that when you are working with HL7 in mirth you are really working with XML, so you could have searched for how to delete a segment.   e.g. https://forums.mirthproject.io/forum/mirth-connect/support/15905-delete-specific-segment

**Reply by JJWest12** [reply]
I did not know that. Will save that link for sure! Thanks again!

**Reply by pacmano1** [reply]
Also read https://svn.wso2.org/repos/wso2/tags/carbon/0.1alpha/mashup/java/xdocs/e4xquickstart.html

**Reply by marctrummer** [reply]
can I ask and additional question as I wasn't able to find a response in the link provided above? I have a similar issue but with repeating nodes. 

```
<Visit>
	<!--more nodes-->
	<Service Transaction="I">
		<!--more nodes-->		
		<ReferrerID>PSY01</ReferrerID>    
	</Service>
</Visit>
<Visit>
	<!--more nodes-->		
	<Service Transaction="I">
		<!--more nodes-->		
		<ReferrerID>PSY02</ReferrerID>    
	</Service>
</Visit>
```

Whenever the ReferrerID inside "Service" is PSY01 I'm trying to delete the whole "Visit" node. The following code is what I tried that obviously didn't work.

```
for each (vis in msg..Visit) {
	for each (serv in vis..Service) {
		if (serv['ReferrerID'] != "PSY01"){
			delete vis; // <--- how do I get this to work?
		}
        }
}
```

**Reply by pacmano1** [reply]
```JS
var xml =
<xml>
<Visit>
	<!--more tags-->
	<Service Transaction="I">
		<!--more tags-->		
		<ReferrerID>PSY01</ReferrerID>    
	</Service>
</Visit>
<Visit>
	<!--more tags-->		
	<Service Transaction="I">
		<!--more tags-->		
		<ReferrerID>PSY02</ReferrerID>    
	</Service>
</Visit>
</xml>

for (var i = xml.length(); i  >= 0 ; i--) {
	logger.info(i)
	if (xml['Visit'][i]['Service']['ReferrerID'].toString() != "PSY01") {
		delete xml['Visit'][i]
	}
}

$c('remaning_xml',xml)
```

**Reply by marctrummer** [reply]
Thanks a lot for the solution. I did not mention that there can be multiple Service nodes within a Visit node. So I just added a second loop to your code and it worked perfectly!
I found another solution using the for each loop I had at first and thought to post it here for anybody that might be interested.

```
for each (vis in msg..Visit) {
	for each (serv in vis..Service) {
		if (serv['ReferrerID'] != "PSY01"){
			delete msg['Visit'][vis.childIndex()];
		}
        }
}
```

**Reply by pacmano1** [reply]
Do read https://forums.mirthproject.io/forum/mirth-connect/support/181472-combine-segments-and-delete-issue

**Reply by marctrummer** [reply]
Thanks a lot for the link! So the loop skips an iteration once a node is deleted?

**Reply by pacmano1** [reply]
Throw some logger.info statements of your full xml before and after the delete and see if your code work as intended (maybe remove the inner loop)

```JS
var xml =
<xml>
<Visit>
	<!--more tags-->
	<Service Transaction="I">
		<!--more tags-->		
		<ReferrerID>PSY01</ReferrerID>    
	</Service>
</Visit>
<Visit>
	<!--more tags-->		
	<Service Transaction="I">
		<!--more tags-->		
		<ReferrerID>PSY02</ReferrerID>    
	</Service>
</Visit>
<Visit>
	<!--more tags-->		
	<Service Transaction="I">
		<!--more tags-->		
		<ReferrerID>PSY03</ReferrerID>    
	</Service>
</Visit>
<Visit>
	<!--more tags-->		
	<Service Transaction="I">
		<!--more tags-->		
		<ReferrerID>PSY04</ReferrerID>    
	</Service>
</Visit>
<Visit>
	<!--more tags-->		
	<Service Transaction="I">
		<!--more tags-->		
		<ReferrerID>PSY01</ReferrerID>    
	</Service>
</Visit>
</xml>

for each (visit in xml.Visit) {	
	logger.info("Before " + xml)
	if (visit['Service']['ReferrerID'].toString() != "PSY01" ) delete visit
	logger.info("After " + xml)
}
```

We would expect only PSY01 to remain, correct?

**Comment by aungktun** [comment]
What is a good reference site for looking up syntax and function/ resources
available for mirth?

**Reply by pacmano1** [reply]
Please don't hijack threads.  Start a new one.
